### PR TITLE
Add octet streaming of sketchs in MSQ

### PR DESCRIPTION
**Pull request by adarshsanjeev**
There are a few issues with using Jackson serialization in sending datasketches between controller and worker in MSQ. This caused a blowup due to holding multiple copies of the sketch being stored. 

This PR aims to resolve this by switching to deserializing the sketch payload without Jackson.

The PR adds a new query parameter used during communication between controller and worker while fetching sketches, "sketchEncoding".
 - If the value of this parameter is OCTET, the sketch is returned as a binary encoding, done by `ClusterByStatisticsSnapshotSerde`. 
- If the value is not the above, the sketch is encoded by Jackson as before.

#### Query param

The worker client now sends a query parameter which specifies the desired format with the request to fetch the sketch. If the worker receives this parameter, it returns the response encoded as a byte array, with the response header "Content-Type":"octet-stream". If the flag is missing, it defaults to the older behavior of Jackson serialized array.

The newer worker client checks the header and deserializes the response accordingly.

If a newer controller is running with an older worker, the worker will ignore the query parameter, and the controller will not see the response header, and use Jackson deserialization. 

If a newer worker is running with an older controller, the query param will not be sent to the worker, which will again default to using Jackson, which the controller will automatically handle.

The above has been tested with an older WorkerClient and newer WorkerChatHandler, and with an older WorkerChatHandler and WorkerClient.

### Byte format

The byte format contains a currently empty header byte, which can be used to store version information. This should allow updating the format in the future, if future versions check the header and deserialize accordingly.

#### Release note
- Reduce memory usage while transferring sketches between MSQ controller and worker.

### Benchmarks

Benchmarks for jackson serde vs octet serde. The octet serde seems to be around 40% faster in most cases.
```
Benchmark                                   (aggregate)  (numBuckets)  (numRows)  Mode  Cnt    Score    Error  Units
BenchmarkSketchTest.benchmarkJacksonSketch         true             1     100000  avgt    5    0.003 ±  0.001  ms/op
BenchmarkSketchTest.benchmarkJacksonSketch         true             1    1000000  avgt    5    0.003 ±  0.001  ms/op
BenchmarkSketchTest.benchmarkJacksonSketch         true          1000     100000  avgt    5    0.990 ±  0.041  ms/op
BenchmarkSketchTest.benchmarkJacksonSketch         true          1000    1000000  avgt    5    0.932 ±  0.046  ms/op
BenchmarkSketchTest.benchmarkJacksonSketch        false             1     100000  avgt    5   18.739 ±  1.410  ms/op
BenchmarkSketchTest.benchmarkJacksonSketch        false             1    1000000  avgt    5   40.421 ±  0.269  ms/op
BenchmarkSketchTest.benchmarkJacksonSketch        false          1000     100000  avgt    5   24.218 ±  0.516  ms/op
BenchmarkSketchTest.benchmarkJacksonSketch        false          1000    1000000  avgt    5  296.199 ± 18.024  ms/op
BenchmarkSketchTest.benchmarkOctetSketch           true             1     100000  avgt    5   ≈ 10⁻³           ms/op
BenchmarkSketchTest.benchmarkOctetSketch           true             1    1000000  avgt    5   ≈ 10⁻³           ms/op
BenchmarkSketchTest.benchmarkOctetSketch           true          1000     100000  avgt    5    0.175 ±  0.002  ms/op
BenchmarkSketchTest.benchmarkOctetSketch           true          1000    1000000  avgt    5    0.193 ±  0.027  ms/op
BenchmarkSketchTest.benchmarkOctetSketch          false             1     100000  avgt    5   10.475 ±  6.384  ms/op
BenchmarkSketchTest.benchmarkOctetSketch          false             1    1000000  avgt    5   24.297 ±  1.731  ms/op
BenchmarkSketchTest.benchmarkOctetSketch          false          1000     100000  avgt    5   14.834 ±  3.089  ms/op
BenchmarkSketchTest.benchmarkOctetSketch          false          1000    1000000  avgt    5  126.198 ± 13.028  ms/op
```
<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/ClusterByStatisticsSnapshotSerde.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  private static final int HEADER_OFFSET = 0;
+  private static final int BUCKET_COUNT_OFFSET = HEADER_OFFSET + Byte.BYTES;
+  private static final int MV_SET_SIZE_OFFSET = BUCKET_COUNT_OFFSET + Integer.BYTES;
+  private static final int MV_VALUES_OFFSET = MV_SET_SIZE_OFFSET + Integer.BYTES;
+
+  private static final int TIMECHUNK_OFFSET = 0;
+  private static final int BUCKET_SIZE_OFFSET = TIMECHUNK_OFFSET + Long.BYTES;
+  private static final int BUCKET_OFFSET = BUCKET_SIZE_OFFSET + Integer.BYTES;
+
+  public static ClusterByStatisticsSnapshot deserialize(ByteBuffer byteBuffer)
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/ClusterByStatisticsSnapshotSerde.java:78]: We should add some benchmarks just to get the relative performance

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/client/SketchResponseHandler.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  @Override
+  public ClientResponse<ClusterByStatisticsSnapshot> done(ClientResponse<BytesFullResponseHolder> response)
+  {
+    return ClientResponse.finished(deserializerFunction.apply(response.getObj()));
+  }
+
+  @Override
+  public void exceptionCaught(ClientResponse<BytesFullResponseHolder> clientResponse, Throwable e)
+  {
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/client/SketchResponseHandler.java:91]: Why would this be empty ?
Shouldn't we do something here ?

**Review by cryptoe — COMMENTED** [review]
Partial review. Will finish the review soon.

**Comment by adarshsanjeev** [comment]
Did some quick benchmarks.

Jackson serialization
----------------------

| Bucket Count | Key Count | Key Size | Serialization Time (ms) | Deserialization Time (ms) |
|--------------|-------------------------------|----------|-------------------------|--------------------------|
| 1            | 148,032                       | 34       |             81            | 46                       |
| 1            | 148,032                     | 100      |              94           | 58                       |
| 1000         | 1,000,000                     | 34       |              193           | 94                       |
| 1000         | 1,000,000                     | 34       |              264           | 258                      |

Octet stream serialization
----------------------

| Bucket Count | Key Count | Key Size | Serialization Time (ms) | Deserialization Time (ms) |
|--------------|-------------------------------|----------|--------------------|--------------------------|
| 1            | 148,032                       | 34       |         17           | 6                        |
| 1            | 148,032                       | 100      |          26          | 11                       |
| 1000         | 1,000,000                     | 34       |          82          | 30                       |
| 1000         | 1,000,000                     | 100      |          144          | 74                       |

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/client/SketchResponseHandler.java (hunk 2 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/client/SketchResponseHandler.java:91]: Similar to org.apache.druid.java.util.http.client.response.BytesFullResponseHandler#exceptionCaught, the response would not be ClientResponse#finished(T). My understanding is that nothing else needs to be done in this case.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/KeyCollectorSnapshotSerializer.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Serializes a {@link ClusterByStatisticsSnapshot} into a byte[].
+ */
+public abstract class KeyCollectorSnapshotSerializer
+{
+  protected abstract byte getType();
+
+  protected abstract byte[] serializeKeyCollector(KeyCollectorSnapshot collectorSnapshot);
+
+  public byte[] serialize(KeyCollectorSnapshot collectorSnapshot)
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/KeyCollectorSnapshotSerializer.java:36]: Please add test cases for null and empty collectors. 
for all the serde methods.

Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/KeyCollectorSnapshotSerializer.java:36]: You should also mention here that type cannot by the same across various implementation. 
Javadocs of the above serialized would be helpful.

**Review by cryptoe — APPROVED** [review]
Left comments. Overall lgtm once the review comments are addressed.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/client/WorkerChatHandler.java [hunk 4]
```diff
@@ -289,4 +309,10 @@ public Response httpGetCounters(@Context final HttpServletRequest req)
     ChatHandlers.authorizationCheck(req, Action.WRITE, task.getDataSource(), toolbox.getAuthorizerMapper());
     return Response.status(Response.Status.OK).entity(worker.getCounters()).build();
   }
+
+  public enum SketchEncoding
+  {
+    OCTET_STREAM,
```
[the remark this answers is not in the document]
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/client/WorkerChatHandler.java:322]: Added

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/ClusterByStatisticsSnapshotSerde.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+{
+  /**
+   * Deserializes the {@link ClusterByStatisticsSnapshot} and writes it to the {@link OutputStream}.
+   * <br>
+   * Format:
+   * - 1 byte : Header byte
+   * - 4 bytes: Number of buckets
+   * - 4 bytes: Number of entries in {@link ClusterByStatisticsSnapshot#getHasMultipleValues()}
+   * - 4 * number of multivalue bucket bytes: List of integers
+   * - number of buckets * buckets, serailized by {@link #serializeBucket(OutputStream, ClusterByStatisticsSnapshot.Bucket)}
```
[the remark this answers is not in the document]
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/ClusterByStatisticsSnapshotSerde.java]: Rewrote this javadoc

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/ClusterByStatisticsSnapshotSerde.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    final byte[] bucketKeyArray = bucket.getBucketKey().array();
+    final double bytesRetained = bucket.getBytesRetained();
+
+    final KeyCollectorSnapshot snapshot = bucket.getKeyCollectorSnapshot();
+    final byte[] serializedSnapshot = snapshot.getSerializer().serialize(snapshot);
+
+    final int length = Double.BYTES + 2 * Integer.BYTES + bucketKeyArray.length + serializedSnapshot.length;
+
+    outputStream.write(
+        ByteBuffer.allocate(Integer.BYTES + length)
```
[the remark this answers is not in the document]
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/ClusterByStatisticsSnapshotSerde.java]: Added a comment

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/KeyCollectorSnapshotSerializer.java (hunk 3 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/KeyCollectorSnapshotSerializer.java:43]: Added test cases for empty collectors. We don't allow null collectors, we throw an error if the request tries to get a non existent collector before this point. Added a notnull annotation instead.
